# Introduction
In this document you will see me go through many emotions.
- [Reinforcement Learning (Cart Pole Game)](#Reinforcement-Learning---Cart-Pole-Game)
- [Reinforcement Learning - Hardcoding](#Reinforcement-Learning---Hardcoding)
- [Custom Reinforcement Learning Algorithm](#Custom-Reinforcement-Learning-Algorithm)
- [My Model vs Random Agent](#My-Model-vs-Random-Agent)

<h1 style="background-color: #9cffb6; text-align:center">Reinforcement Learning - Cart Pole Game</h1>

So in this section I TRIED to get an RL agent working for cart pole game. but i just couldnt figure it out. Here's some code I found online on [this website](https://pylessons.com/CartPole-reinforcement-learning). I did modify the code like a little bit but nothing that big. I mainly understood what was going on and that's something right?

## Install dependencies

In [1]:
!pip install gym
!pip install gym[box2d]

## Import Dependencies

In [1]:
import gym
print("gym version: " + gym.__version__)
import random

gym version: 0.26.2


## Coding

In [9]:
env = gym.make("CartPole-v1")
env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

MAX_AMOUNT_OF_TRIES = 100
MAX_AMOUNT_OF_FRAMES_PER_TRY = 500

def Random_games():
    # a try at the game
    for episode in range(MAX_AMOUNT_OF_TRIES):
        # reset all previous made progress in the environment
        env.reset()
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_TRY):
            # make the environment visible
            env.render()
            
            # generate random action (0=left or 1=right)
            action = env.action_space.sample()

            # output of the action
            next_state, reward, done, info, _ = env.step(action)
            
            # print results
            print(t, next_state, reward, done, info, action)
            if done:
                break
                
Random_games()

0 [ 0.04637115  0.22441569  0.04807265 -0.29248196] 1.0 False False 1
1 [ 0.05085946  0.41882044  0.04222301 -0.56962425] 1.0 False False 1
2 [ 0.05923587  0.61332554  0.03083053 -0.848712  ] 1.0 False False 1
3 [ 0.07150238  0.80801374  0.01385629 -1.1315428 ] 1.0 False False 1
4 [ 0.08766266  1.0029516  -0.00877457 -1.4198478 ] 1.0 False False 1
5 [ 0.10772169  1.198181   -0.03717153 -1.7152605 ] 1.0 False False 1
6 [ 0.1316853   1.3937092  -0.07147674 -2.019276  ] 1.0 False False 1
7 [ 0.15955949  1.5894955  -0.11186226 -2.333202  ] 1.0 False False 1
8 [ 0.1913494  1.7854369 -0.1585263 -2.6580958] 1.0 False False 1
9 [ 0.22705814  1.981349   -0.21168822 -2.9946911 ] 1.0 True False 1
0 [-0.02236718  0.18103437  0.01605817 -0.28207645] 1.0 False False 1
1 [-0.01874649  0.37592363  0.01041664 -0.5696517 ] 1.0 False False 1
2 [-0.01122802  0.57089794 -0.00097639 -0.8590349 ] 1.0 False False 1
3 [ 1.8993927e-04  7.6603317e-01 -1.8157093e-02 -1.1520247e+00] 1.0 False False 1
4 [ 0.0155106

5 [ 0.03086281  1.121255   -0.04543807 -1.6962374 ] 1.0 False False 1
6 [ 0.05328791  1.3168707  -0.07936282 -2.0027122 ] 1.0 False False 1
7 [ 0.07962533  1.5127258  -0.11941707 -2.3188784 ] 1.0 False False 1
8 [ 0.10987984  1.7087157  -0.16579464 -2.6457946 ] 1.0 False False 1
9 [ 0.14405416  1.9046544  -0.21871053 -2.984193  ] 1.0 True False 1
0 [ 0.0104518   0.21349826 -0.03382784 -0.26310664] 1.0 False False 1
1 [ 0.01472177  0.40908632 -0.03908997 -0.5662644 ] 1.0 False False 1
2 [ 0.0229035   0.60473424 -0.05041526 -0.8710016 ] 1.0 False False 1
3 [ 0.03499818  0.8005043  -0.06783529 -1.1790994 ] 1.0 False False 1
4 [ 0.05100827  0.9964383  -0.09141728 -1.4922527 ] 1.0 False False 1
5 [ 0.07093704  1.1925461  -0.12126233 -1.8120254 ] 1.0 False False 1
6 [ 0.09478796  1.3887925  -0.15750284 -2.1397972 ] 1.0 False False 1
7 [ 0.12256381  1.5850822  -0.20029879 -2.4767017 ] 1.0 False False 1
8 [ 0.15426545  1.7812415  -0.24983282 -2.823553  ] 1.0 True False 1
0 [ 0.02506913  0.1919

8 [ 0.15436457  1.7522058  -0.17654467 -2.721726  ] 1.0 False False 1
9 [ 0.18940869  1.948096   -0.23097919 -3.062629  ] 1.0 True False 1
0 [ 0.04788985  0.17175671 -0.02675902 -0.28647524] 1.0 False False 1
1 [ 0.05132499  0.36724988 -0.03248853 -0.5874762 ] 1.0 False False 1
2 [ 0.05866999  0.5628114  -0.04423805 -0.89021367] 1.0 False False 1
3 [ 0.06992622  0.75850475 -0.06204233 -1.1964682 ] 1.0 False False 1
4 [ 0.08509631  0.9543725  -0.08597169 -1.5079334 ] 1.0 False False 1
5 [ 0.10418376  1.1504251  -0.11613036 -1.8261707 ] 1.0 False False 1
6 [ 0.12719226  1.3466278  -0.15265377 -2.152558  ] 1.0 False False 1
7 [ 0.15412483  1.5428858  -0.19570494 -2.4882278 ] 1.0 False False 1
8 [ 0.18498254  1.7390261  -0.24546948 -2.8339925 ] 1.0 True False 1
0 [ 0.00428543  0.205483    0.01503801 -0.25085214] 1.0 False False 1
1 [ 0.00839509  0.40038702  0.01002097 -0.5387541 ] 1.0 False False 1
2 [ 1.6402829e-02  5.9536666e-01 -7.5411162e-04 -8.2826275e-01] 1.0 False False 1
3 [ 0.0283

6 [ 0.06835044  1.3485044  -0.09454019 -2.047938  ] 1.0 False False 1
7 [ 0.09532053  1.54446    -0.13549894 -2.3683124 ] 1.0 False False 1
8 [ 0.12620972  1.7405007  -0.18286519 -2.6993928 ] 1.0 False False 1
9 [ 0.16101974  1.9364202  -0.23685305 -3.0418355 ] 1.0 True False 1
0 [-0.00466221  0.16441023 -0.04907552 -0.31320158] 1.0 False False 1
1 [-0.001374    0.3601957  -0.05533955 -0.6209486 ] 1.0 False False 1
2 [ 0.00582991  0.55604506 -0.06775852 -0.9305344 ] 1.0 False False 1
3 [ 0.01695081  0.75201285 -0.08636921 -1.2437173 ] 1.0 False False 1
4 [ 0.03199107  0.94813037 -0.11124356 -1.5621581 ] 1.0 False False 1
5 [ 0.05095368  1.1443933  -0.14248672 -1.887371  ] 1.0 False False 1
6 [ 0.07384154  1.340748   -0.18023413 -2.2206676 ] 1.0 False False 1
7 [ 0.1006565   1.537075   -0.22464749 -2.5630903 ] 1.0 True False 1
0 [ 0.02209791  0.22808601  0.00839556 -0.305779  ] 1.0 False False 1
1 [ 0.02665963  0.42308733  0.00227998 -0.5958024 ] 1.0 False False 1
2 [ 0.03512138  0.6181

KeyboardInterrupt: 

# Conclusion
It might look like I gave up quickly but. Trust me I tried a lot. I just can't make it work. 

<h1 style="background-color: #9cffb6; text-align:center">Reinforcement Learning - Hardcoding</h1>

I was curious if I could maybe hardcode a solution. With this code we move left if pole falls left and right if pole falls right

- Before I start with all of this. I want to set a goal for myself. Let's say we aim for a score of 500. That means we live 500 frames.

In [77]:
#env = gym.make("CartPole-v1")
env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

MAX_AMOUNT_OF_TRIES = 100
MAX_AMOUNT_OF_FRAMES_PER_TRY = 500

def Random_games():
    # a try at the game
    for episode in range(MAX_AMOUNT_OF_TRIES):
        # reset all previous made progress in the environment
        env.reset()
        
        score = 0
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_TRY):
            # make the environment visible
            env.render()
            
            # set action, if pole falls right move right vice versa
            if env.state[2] > 0:
                action = 1
            else:
                action = 0
            
            # output of the action
            next_state, reward, done, info, _ = env.step(action)
            
            # keep track of score
            score+=1
            if done:
                print('score: ' + str(score))
                break
                
Random_games()

score: 40
score: 47
score: 34
score: 49
score: 53
score: 46
score: 56
score: 66
score: 35
score: 50
score: 53
score: 47
score: 41
score: 24
score: 47
score: 35
score: 38
score: 44
score: 52
score: 39
score: 35
score: 61
score: 25
score: 53
score: 49
score: 39
score: 66
score: 25
score: 47
score: 36
score: 39
score: 35
score: 38
score: 54
score: 36
score: 52
score: 55
score: 35
score: 38
score: 51
score: 34
score: 35
score: 52
score: 26
score: 41
score: 31
score: 62
score: 38
score: 41
score: 45
score: 35
score: 39
score: 25
score: 39
score: 40
score: 25
score: 37
score: 40
score: 27
score: 35
score: 47
score: 47
score: 40
score: 46
score: 52
score: 47
score: 40
score: 48
score: 58
score: 40
score: 44
score: 45
score: 38
score: 41
score: 53
score: 39
score: 41
score: 46
score: 31
score: 53
score: 40
score: 36
score: 45
score: 40
score: 37
score: 42
score: 41
score: 45
score: 54
score: 46
score: 40
score: 39
score: 41
score: 40
score: 51
score: 34
score: 31
score: 53
score: 63
score: 36


# Conclusion
This wasn't that much better. our average moved from 20-30 to 40-50. But I want 500. So we need something better.

<h1 style="background-color: #9cffb6; text-align:center">Custom Reinforcement Learning Algorithm</h1>

## Introduction
So I understand the concept of Reinforcement Learning, but I just can't seem to make the code from the internet work. Everything is always deprecated or uses different versions that aren't available, getting errors all day. So I thought by myself. If I really understand it, I should be able to make it myself right?

## Train with rewards
So, we start of by receiving rewards. How can we define a reward in this simplistic game? My idea was to give a higher reward the closer the pole speed gets to 0. Because the slower the pole moves, the more balanced it is.

My reward System
```
pole_speed: 0 - 0.05 = reward 6
pole_speed: 0.05 - 0.1 = reward 5
pole_speed: 0.1 - 0.2 = reward 4
pole_speed: 0.2 - 0.3 = reward 3
pole_speed: 0.3 - 0.4 = reward 2
pole_speed: 0.4 - 0.5 = reward 1
pole_speed: >0.5 = reward 0
```
(speed is the 4th variable in the state array in the environment)

## possible problem for later
this will probably go very slow but I don't mind really lets just do cool things

## Data Sourcing
Here's the idea to collect data
1. we have the state of the game
2. we do a random move
3. we get a new state
4. look at the new state and give a reward based on the new state
5. save the old state, action and reward all together

We should have something like this in the end
`brain_array = [state, action, reward]`

So whilst training we keep adding more entries to brain array. We can even overwrite an exact copy of an entry when the reward is better than the reward previous saved. Sounds complicated, let me show you what i mean.

lets say we have this data
```
brain_array = [
1. [[0, 0, 1, 0], "right", 1]
2. [[0, 1, 0, 0], "right", 1]
3. [[1, 0, 1, 0], "left", 3]
]
```
we get a new `entry` whilst training, and this new entry has the same `state` as entry 1 but a better `reward`.
`new_entry = [[0, 0, 1, 0], "left", 5]`

in this case we will overwrite the entry in brain_array. (entry 1 has now been overwritten with a better move)
```
brain_array = [
1. [[0, 0, 1, 0], "left", 5]
2. [[0, 1, 0, 0], "right", 1]
3. [[1, 0, 1, 0], "left", 3]
]
```

#### lets go

# Trying to execute everything i explained

here we declare `brain_array`. It's the array where we save all data in. The idea is to later on per frame check if `brain_array` has already experienced this game state before. If it did happen before, brain_array can return an action for my model to take.

In [226]:
# declare / reset brain_array # we fill the brain array with a irrelevant value since it needs at least one value to not crash later on
brain_array = [[[0.00680, 0.00819, 0.02738, -0.01811], 0, 0]]

with the `add_brain_cell` function we can add data to `brain_array`. it also automatically overwrites an entry with the same state that has a better result

**DECIMALS**: the decimal variable is major factor in our learning algorithm. The state actually looks something like this `[0.05365345720, 0.00813546346579, 0.027334573468, -0.01811345735673]`. If we want to test my model, we will check for every frame if the current state has happened before. But if there are 9 decimals this is incredibly specific. With 9 decimals there are a total of 620.000.000.000.000.000.000.000.000.000.000.000.000.000 possible states. So you will never get the same state. That's why we decrease the Decimals for higher chance of encountering a state we have already seen before. 

**5 Decimals**: I tried reducing to 5 decimals but with 15k+ unique states in the brain_array, we never encountered the same state twice. 

**3 Decimals**: Still no duplicate state.

**2 Decimals**: After putting decimals on just 2, we already have a way higher chance of finding states we encountered before. Because 2 decimals leaves us with a total of 25.920.000.000 states that's already way better. Also, we will find the same states more than you think just because some possible states you will never be in. like the cart position is mostly just in the middle of the map leaving out already so much possibilities. I have tried 3 models on 2 decimals, I will show results later.

**1 Decimal**: 1 decimal has only 720.000 possible states. If lots of the possible states are irrelevant. It means we will almost find the same state every frame. So 1 Decimal is definitely the quickest way of training.

In [359]:
# the lower decimals the less precice the agent gets. but lowering this... 
# ...number does speed up training requiring less data so its something you have to weigh a decision of
DECIMALS = 1

def add_brain_cell(state, action, reward):
    # declare i so we can keep track of the current item in the loop and modify it
    i = 0
    for entry in brain_array:
        # for some reason the state is not an array so we make it an array first
        # HEY ITS FUTURE ME, i came back here to round off all numbers to 5 decimals you'll find out later why
        state_but_array = [round(state[0], DECIMALS), round(state[1], DECIMALS), round(state[2], DECIMALS), round(state[3], DECIMALS)]
        if entry[0] == state_but_array:
            #print('same state')
            if reward > entry[2]:
                print('better reward')
                brain_array[i] = [state_but_array, action, reward]
                return True 
            else:
                break
        i+=1
    brain_array.append([state_but_array, action, reward])
                
            
            

## Calculating the reward
code for calculating the reward, read the comments for more info, Here we only use the pole speed as input for the reward.

In [83]:
def calc_reward(pole_velocity):
    
    # declare reward
    reward = 0
    
    # since were dealing with -/+ values we run abs to make it an absolute value. because we dont care if its + or -
    pole_velocity = abs(pole_velocity)
    
    # check what the reward should be and return it
    if pole_velocity <= 0.05:
        reward = 6
    elif pole_velocity <= 0.1:
        reward = 5
    elif pole_velocity <= 0.2:
        reward = 4
    elif pole_velocity <= 0.3:
        reward = 3
    elif pole_velocity <= 0.4:
        reward = 2
    elif pole_velocity <= 0.5:
        reward = 1
    return reward
        
        
        

## Add a brain cell (1 data entry)
if you run this code once, it will add 1 entry of data to `brain_array` How amazing! (I also used this as a trial and error place to test the functions i wrote)

In [12]:
env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

MAX_AMOUNT_OF_TRIES = 1
MAX_AMOUNT_OF_FRAMES_PER_TRY = 500

def Random_games():
    for episode in range(MAX_AMOUNT_OF_TRIES):
        env.reset()
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_TRY):
            env.render()
            
            # choose random action | choose action based of earlier done moves 
            action = random.choice([0, 1]) # we use random lib for quicker performance

            # get the current stat so we can save it later as data
            old_state = env.state                      
            # proceed a frame in the game (rhyms 😎)
            env.step(action)
            
            # use the current state values to calulate the reward so we can decide later how good the move was
            reward = calc_reward(env.state[3])
            
            # give values myself for testing
            old_state, action, reward = [[-0.083, -0.81, 0.098, 1.262], 0, 1]
            add_brain_cell(old_state, action, reward)
            
            print(brain_array)
            break
            
            # print results
            print(next_state[3])
            if done:
                break
                
Random_games()

[[[0.0068, 0.00819, 0.02738, -0.01811], 0, 0], [[-0.08, -0.81, 0.1, 1.26], 0, 1]]


# COLLECT A LOT OF DATA (BRAIN MIGHT EXPLODE🧠💥)

In this section we purely focus on collecting data

In [19]:
env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

MAX_AMOUNT_OF_TRIES = 1000
MAX_AMOUNT_OF_FRAMES_PER_TRY = 500

def Random_games():
    for episode in range(MAX_AMOUNT_OF_TRIES):
        env.reset()
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_TRY):
            env.render()
            
            # choose random action | choose action based of earlier done moves 
            action = random.choice([0, 1]) # we use random lib for quicker performance

            # get the current stat so we can save it later as data
            old_state = env.state
                                   
            # proceed a frame in the game (rhyms 😎)
            env.step(action)
            
            # use the current state values to calulate the reward so we can decide later how good the move was
            reward = calc_reward(env.state[3])
            
            # add values
            add_brain_cell(old_state, action, reward)
            
            # only get done to check if the game is done
            _, _, done, _, _ = env.step(action)
            if done:
                break
                
Random_games()

same state
better reward
same state
same state
better reward
same state
same state
same state
same state
same state
same state
same state
same state
better reward
same state
same state
better reward
same state
same state
better reward
same state
same state
same state
same state
same state
better reward
same state
same state
same state
better reward
same state
same state
better reward
same state
same state
same state
same state
same state
same state
same state
same state
better reward
same state
same state
same state
same state
better reward
same state
better reward
same state
same state
better reward
same state
same state
same state
same state
same state
same state
better reward
same state
same state
same state
same state
same state
same state
same state
better reward
same state
same state
better reward
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
better reward
same s

KeyboardInterrupt: 

You can see `same state` output meaning this state has been encountered before. If you also see the `better reward` output it means the action in brain_array has been overwritten with a better action (because it gives a better reward)

<h1 style="background-color: #9cffb6; text-align:center">My Model vs Random Agent</h1>

so i have trained a model with 2 decimals. It has trained a little over 10000 frames, 15000 frames and 375000 frames. Let's put it to the test and see if it will perform better than the random agent

Here we declare a function to calculate what the average score is we get so we can compare the two.

In [40]:
def get_avg_score():
    total = 0
    for x in scores_array:
        total+=x
    return total/MAX_AMOUNT_OF_TRIES

In [49]:
env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_TRIES = 100
MAX_AMOUNT_OF_FRAMES_PER_TRY = 500

def Random_games():
    # a try at the game
    for episode in range(MAX_AMOUNT_OF_TRIES):
        # reset all previous made progress in the environment
        env.reset()
        
        score = 0
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_TRY):
            # make the environment visible
            env.render()
            
            # set action, if pole falls right move right vice versa
            action = env.action_space.sample()
            
            # output of the action
            next_state, reward, done, info, _ = env.step(action)
            
            # keep track of score
            score+=1
            if done:
                print('score: ' + str(score))
                scores_array.append(score)
                break

Random_games()
print('average score: ' + str(get_avg_score()))

score: 25
score: 11
score: 50
score: 16
score: 23
score: 21
score: 14
score: 25
score: 27
score: 45
score: 18
score: 13
score: 13
score: 19
score: 34
score: 12
score: 17
score: 10
score: 9
score: 37
score: 42
score: 67
score: 12
score: 17
score: 28
score: 26
score: 13
score: 15
score: 13
score: 33
score: 12
score: 13
score: 16
score: 26
score: 20
score: 31
score: 16
score: 11
score: 13
score: 15
score: 14
score: 17
score: 10
score: 13
score: 14
score: 12
score: 80
score: 12
score: 49
score: 14
score: 24
score: 19
score: 42
score: 15
score: 12
score: 15
score: 18
score: 40
score: 18
score: 26
score: 15
score: 34
score: 17
score: 17
score: 9
score: 14
score: 15
score: 27
score: 20
score: 11
score: 17
score: 17
score: 15
score: 25
score: 15
score: 52
score: 24
score: 25
score: 25
score: 60
score: 12
score: 12
score: 33
score: 42
score: 83
score: 16
score: 59
score: 19
score: 16
score: 12
score: 14
score: 22
score: 29
score: 30
score: 29
score: 22
score: 16
score: 36
score: 28
score: 14
av

random agent scores 23.4

## Get best action from brain_array
last thing before we start, we need a function to get the best action. this basically looks in the model brain and looks if the current state has already been solved before, if the result is bigger than 0 it will do the provided action. if its 0 it will do the other action.

**TRY_IMPROVE_REWARD**: this basically means, if the reward currently is for example 3 than we will try another action to see if we can improve 3. The higher this number the bigger the chance of getting better actions, but also the bigger the chance of messing up the game since it might go for worse actions as well. So this variable is my own `Exploration` variable you usually see in RL.

In [250]:
# you can raise the value of this variable to try to improve existing states. the higher the number, the better the chance you will improve an older action. But if you want to go for a highscore might want to consider to turn it off since it can increase the chances on mistakes aswell. You can see it as an exploration variable
TRY_IMPROVE_REWARD = 5 # (bascially the exploration number can be 0-5)

def get_action(state):
    state_but_array = [round(state[0], DECIMALS), round(state[1], DECIMALS), round(state[2], DECIMALS), round(state[3], DECIMALS)]
    for entry in brain_array:
        if entry[0] == state_but_array:
            #print('found state')
            if entry[2] > TRY_IMPROVE_REWARD:
                return entry[1]
            else:
                if entry[1] == 0:
                    return 1
                else:
                    return 0
    return random.choice([0, 1])

here we declare a function to load a model

In [222]:
import json

def get_model(model):
    with open(model + '.json', 'r') as f:
        data = json.load(f)
        return data

and here we declare a function to extract a model from the program

In [251]:
def save_model(model, data):
    save_file = open(model + '.json', "w")  
    json.dump(data, save_file, indent = 6)  
    save_file.close()  

Current models and the amount of trained frames

2 Decimal Models
- 2dec-model-v1 (10k)
- 2dec-model-v2 (15k)
- 2dec-model-v3 (37.5k)

1 Decimal Models
- 1dec-model-v1 (15k)

The sad part is that I saved all of these in .txt format so they are all gone now 🤠👍

Let's create a new 1 decimal model

# Creating a 1 decimal model
Let's first reset brain_array

In [239]:
brain_array = [[[0.0, 0.0, 0.0, -0.0], 0, 0]]

In [249]:
# first train once, after training once uncomment this to keep learning from past performance
#brain_array=get_model('1dec-model-v2')

env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_TRIES = 100
MAX_AMOUNT_OF_FRAMES_PER_TRY = 500

def Train_and_test():
    # a try at the game
    for episode in range(MAX_AMOUNT_OF_TRIES):
        # reset all previous made progress in the environment
        env.reset()
        
        score = 0
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_TRY):
            # make the environment visible
            #env.render()
            
            # set action, if pole falls right move right vice versa
            action = get_action(env.state)
            
            
            
            # get the current stat so we can save it later as data
            old_state = env.state
            
            
            
            # output of the action
            _, _, done, _, _ = env.step(action)
            
            
            
            # use the current state values to calulate the reward so we can decide later how good the move was
            reward = calc_reward(env.state[3])
            
            # add values
            add_brain_cell(old_state, action, reward)
            
            
            
            # keep track of score
            score+=1
            if done:
                print('score: ' + str(score))
                scores_array.append(score)
                break

Train_and_test()
print('average score: ' + str(get_avg_score()))
save_model('1dec-model-v2', brain_array)

same state
same state
same state
same state
same state
same state
same state
same state
same state
better reward
same state
same state
better reward
same state
same state
better reward
score: 24
same state
same state
same state
same state
same state
same state
same state
same state
score: 15
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 22
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
better reward
same state
better reward
same state
same state
same state
same state
better reward
same state
better reward
same state
same state
same state
same state
better reward
same state
better reward
same state
same state
better reward
score: 54
same state
same state
same state
same state
same state
same state
same state
same s

same state
same state
same state
score: 29
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 11
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 15
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 13
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 23
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 11
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 9
same state
same state
same state
same st

same state
score: 15
same state
same state
same state
same state
same state
same state
same state
same state
score: 11
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 12
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 13
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 16
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
better reward
same state
same state
better reward
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
score: 34
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state
same state

In [252]:
brain_array = get_model('1dec-model-v2')
print(len(brain_array))

3050


# Time to test
Now we have a 1 decimal model with a 3k frame experience. lets see if we get better results than the random agent

In [284]:
brain_array=get_model('1dec-model-v2')

env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_GAMES = 100
MAX_AMOUNT_OF_FRAMES_PER_GAME = 500

def Test():
    for episode in range(MAX_AMOUNT_OF_GAMES):
        env.reset()
        score = 0
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_GAME):
            #env.render()
            
            # get action
            action = get_action(env.state)    
            
            # output of the action
            _, _, done, _, _ = env.step(action)
            
            # keep track of score
            score+=1
            if done:
                print('score: ' + str(score))
                scores_array.append(score)
                break

Test()
print('average score: ' + str(get_avg_score()))

score: 52
score: 39
score: 28
score: 51
score: 73
score: 33
score: 61
score: 41
score: 128
score: 60
score: 49
score: 43
score: 44
score: 51
score: 58
score: 35
score: 78
score: 56
score: 31
score: 66
score: 48
score: 76
score: 28
score: 37
score: 35
score: 15
score: 13
score: 50
score: 76
score: 17
score: 97
score: 60
score: 48
score: 64
score: 50
score: 51
score: 44
score: 62
score: 34
score: 43
score: 49
score: 39
score: 38
score: 38
score: 53
score: 100
score: 58
score: 103
score: 46
score: 13
score: 65
score: 112
score: 41
score: 56
score: 65
score: 78
score: 40
score: 65
score: 55
score: 42
score: 112
score: 40
score: 88
score: 60
score: 75
score: 98
score: 53
score: 47
score: 73
score: 44
score: 34
score: 36
score: 40
score: 36
score: 33
score: 56
score: 105
score: 68
score: 27
score: 99
score: 37
score: 95
score: 11
score: 60
score: 35
score: 63
score: 59
score: 38
score: 24
score: 53
score: 25
score: 51
score: 67
score: 118
score: 56
score: 56
score: 65
score: 72
score: 53
sco

Our score is even lower than the random agent, Let's train more and go again.

Scoreboard
1. Random Agent: 23.4
2. 1dec-model-v2: 19.08  (3050)

# Using and Training model at the same time
Let's make the code shorter by removing all comments

In [310]:
env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

brain_array = get_model('1dec-model-v3')
scores_array = []
MAX_AMOUNT_OF_GAMES = 1
MAX_AMOUNT_OF_FRAMES_PER_GAME = 500
TRY_IMPROVE_REWARD = 1 # exploration variable

def get_action(state):
    state_but_array = [round(state[0], DECIMALS), round(state[1], DECIMALS), round(state[2], DECIMALS), round(state[3], DECIMALS)]
    for entry in brain_array:
        if entry[0] == state_but_array:
            if entry[2] > TRY_IMPROVE_REWARD:
                return entry[1]
            else:
                if entry[1] == 0:
                    return 1
                else:
                    return 0
    return random.choice([0, 1])

def get_avg_score(MAX_AMOUNT_OF_GAMES):
    total = 0
    for x in scores_array:
        total+=x
    return total/MAX_AMOUNT_OF_GAMES

In [324]:
def Train():
    for episode in range(MAX_AMOUNT_OF_GAMES):
        env.reset()
        score = 0
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_GAME):
            #env.render()
            action = get_action(env.state)
            old_state = env.state
            _, _, done, _, _ = env.step(action)
            reward = calc_reward(env.state[3])
            add_brain_cell(old_state, action, reward)
            
            # keep track of score
            score+=1
            if done:
                print('score: ' + str(score))
                scores_array.append(score)
                break
Train()
print('average score: ' + str(get_avg_score(MAX_AMOUNT_OF_GAMES)))
save_model('1dec-model-v3', brain_array)

UnboundLocalError: local variable 'pole_speed' referenced before assignment

In [281]:
brain_array = get_model('1dec-model-v3')
print(len(brain_array))

11846


In [288]:
brain_array=get_model('1dec-model-v3')

env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_GAMES = 100
MAX_AMOUNT_OF_FRAMES_PER_Game = 500
TRY_IMPROVE_REWARD = 0 # exploration variable

Test()
print('average score: ' + str(get_avg_score()))

score: 94
score: 80
score: 141
score: 83
score: 85
score: 108
score: 149
score: 114
score: 89
score: 126
score: 98
score: 131
score: 128
score: 106
score: 102
score: 108
score: 91
score: 88
score: 118
score: 74
score: 117
score: 102
score: 92
score: 109
score: 54
score: 68
score: 122
score: 85
score: 89
score: 136
score: 103
score: 150
score: 108
score: 42
score: 98
score: 133
score: 101
score: 64
score: 129
score: 93
score: 93
score: 76
score: 72
score: 89
score: 103
score: 139
score: 113
score: 157
score: 77
score: 57
score: 67
score: 133
score: 81
score: 94
score: 80
score: 129
score: 79
score: 170
score: 90
score: 81
score: 82
score: 71
score: 55
score: 91
score: 110
score: 81
score: 74
score: 113
score: 125
score: 85
score: 129
score: 94
score: 64
score: 87
score: 122
score: 79
score: 108
score: 103
score: 142
score: 90
score: 86
score: 96
score: 67
score: 75
score: 131
score: 65
score: 99
score: 60
score: 117
score: 126
score: 68
score: 118
score: 87
score: 99
score: 238
score: 1

This is great, we start to see progression. The results aren't random anymore.

Scoreboard
1. 1dec-model-v3: 99.89 (11846)
2. Random Agent: 23.4
3. 1dec-model-v2: 19.08  (3050)

# Reward Optimalisation
I have observed the gameplay of my model and saw room for improvement. I teached my model how to keep the pole speed as close to 0 as possible. There is one small problem here. If the pole leans to a direction, the speed can still be held close to 0, so the agent keeps the speed at 0 but the pole is slowly falling. to solve this, we can make a new reward system where we can take into account both pole angle and pole speed, but first, lets try only pole angle.

In [331]:
def calc_reward(pole_angle):
    
    # declare reward
    reward = 0
    
    # since were dealing with -/+ values we run abs to make it an absolute value. because we dont care if its + or -
    pole_angle = abs(pole_angle)
    
    # check what the reward should be and return it
    if pole_angle <= 0.1:
        reward = 5
    elif pole_angle <= 0.2:
        reward = 4
    return reward

since we will use a different method we will need a new fresh model. so we're resetting brain_array

In [301]:
brain_array = [[[0.0,0.0,0.0,0.0], 0, 0]] # reset

In [321]:
env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_GAMES = 100
MAX_AMOUNT_OF_FRAMES_PER_GAME = 500
Test()
print('average score: ' + str(get_avg_score(MAX_AMOUNT_OF_GAMES)))
save_model('1dec-model-angle-v1', brain_array)

score: 104
score: 140
score: 75
score: 137
score: 92
score: 145
score: 68
score: 114
score: 101
score: 111
score: 95
score: 43
score: 132
score: 107
score: 127
score: 121
score: 81
score: 99
score: 116
score: 124
score: 78
score: 75
score: 72
score: 56
score: 134
score: 117
score: 142
score: 67
score: 101
score: 90
score: 114
score: 73
score: 47
score: 96
score: 97
score: 139
score: 121
score: 110
score: 102
score: 66
score: 52
score: 199
score: 89
score: 123
score: 100
score: 85
score: 69
score: 147
score: 71
score: 107
score: 86
score: 84
score: 84
score: 88
score: 153
score: 140
score: 143
score: 127
score: 155
score: 96
score: 100
score: 141
score: 105
score: 118
score: 250
score: 110
score: 99
score: 179
score: 84
score: 123
score: 172
score: 78
score: 108
score: 123
score: 113
score: 140
score: 108
score: 96
score: 128
score: 87
score: 111
score: 44
score: 143
score: 111
score: 124
score: 107
score: 170
score: 107
score: 68
score: 108
score: 138
score: 122
score: 136
score: 49
sc

so this works well, but it brings another problem. Now the cart just slowly moves to a side and goes against the side of the screen. This isn't good since we can't move left infinitely. To fix this we could take into account the cart location aswel. But before we do that, let's see if we can get 500 with pole angle + pole speed

Scoreboard
1. 1dec-model-angle-v1: 107.00 (13121) (angle method)
2. 1dec-model-v3: 99.89 (11846) (speed method)
3. Random Agent: 23.4
4. 1dec-model-v2: 19.08  (3050) (speed method)

In [349]:
def calc_reward_both(pole_speed, pole_angle):
    
    # declare reward
    reward = 0
    
    # since were dealing with -/+ values we run abs to make it an absolute value. because we dont care if its + or -
    pole_speed = abs(pole_speed)
    pole_angle = abs(pole_angle)
    
    # check what the reward should be and return it
    if pole_speed <= 0.1 and pole_angle <= 0.1:
        reward = 5
    elif pole_speed <= 0.2 and  pole_angle <= 0.1:
        reward = 4
    elif pole_speed <= 0.3 and pole_angle <= 0.1:
        reward = 3
    elif pole_speed <= 0.2 and pole_angle <= 0.2:
        reward = 2
    elif pole_speed <= 0.3 and pole_angle <= 0.2:
        reward = 1
    return reward

In [ ]:
def get_avg_score(score_array, max_games):
    total = 0
    for x in score_array:
        total+=x
    return total/max_games

we reset brain_array because we have a new method

In [339]:
brain_array = [[[0.0,0.0,0.0,0.0], 0, 0]] # reset

In [377]:
brain_array = get_model('1dec-model-both-v1')

env = gym.make("CartPole-v1")
#env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_GAMES = 10000
MAX_AMOUNT_OF_FRAMES_PER_GAME = 500
TRY_IMPROVE_REWARD = 0 # exploration variable

def Train():
    for episode in range(MAX_AMOUNT_OF_GAMES):
        env.reset()
        score = 0
        for t in range(MAX_AMOUNT_OF_FRAMES_PER_GAME):
            #env.render()
            action = get_action(env.state)
            old_state = env.state
            _, _, done, _, _ = env.step(action)
            reward = calc_reward_both(env.state[3], env.state[2])
            add_brain_cell(old_state, action, reward)
            
            # keep track of score
            score+=1
            if done:
                print('score: ' + str(score))
                scores_array.append(score)
                break
    save_model('1dec-model-both-v1', brain_array)

Train()
print('average score: ' + str(get_avg_score(MAX_AMOUNT_OF_GAMES)))
save_model('1dec-model-both-v1', brain_array)

score: 18
better reward
score: 70
better reward
better reward
better reward
better reward
better reward
score: 73
better reward
better reward
better reward
better reward
score: 91
better reward
better reward
score: 92
better reward
score: 60
better reward
score: 60
score: 91
better reward
better reward
better reward
better reward
better reward
better reward
score: 145
score: 19
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
score: 159
score: 188
score: 33
better reward
better reward
score: 102
better reward
better reward
better reward
better reward
better reward
score: 118
better reward
better reward
better reward
better reward
better reward
better reward
score: 213
better reward
better reward
better reward
score: 137
better reward
better reward
better reward
better reward
better reward
better reward
better reward
score: 118
score: 45
better reward
better reward
better reward
better reward
better reward
scor

score: 145
score: 90
score: 94
score: 90
better reward
better reward
better reward
better reward
better reward
better reward
better reward
score: 136
better reward
score: 79
score: 108
score: 27
score: 62
score: 87
score: 147
score: 59
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
score: 126
score: 125
better reward
score: 167
score: 93
score: 43
score: 23
better reward
score: 141
score: 131
score: 88
score: 164
better reward
score: 61
better reward
better reward
score: 97
score: 105
better reward
better reward
better reward
score: 116
score: 101
score: 131
score: 126
score: 46
better reward
score: 110
better reward
better reward
score: 120
score: 102
score: 80
score: 151
better reward
better reward
better reward
better reward
better reward
better reward
score: 198
score: 123
better reward
score: 104
score: 94
better reward
score: 151
score: 88
better reward
better reward
better reward
better reward
better reward
better 

score: 113
better reward
score: 117
score: 89
score: 18
better reward
better reward
better reward
better reward
better reward
better reward
score: 180
better reward
better reward
better reward
better reward
score: 148
better reward
better reward
better reward
score: 161
better reward
better reward
better reward
score: 159
better reward
score: 129
score: 68
better reward
better reward
better reward
score: 142
score: 53
score: 128
score: 57
score: 43
better reward
score: 137
score: 89
score: 102
score: 17
better reward
better reward
score: 117
score: 63
score: 119
score: 99
score: 17
score: 268
score: 169
score: 95
score: 101
better reward
better reward
better reward
better reward
score: 219
score: 78
score: 23
score: 120
score: 106
score: 81
score: 115
score: 127
better reward
better reward
score: 109
better reward
score: 201
score: 75
score: 68
better reward
score: 128
better reward
better reward
better reward
better reward
score: 141
better reward
better reward
better reward
score: 18

score: 150
better reward
better reward
better reward
score: 167
score: 33
score: 98
score: 45
score: 19
score: 88
better reward
score: 79
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
better reward
score: 198
score: 88
score: 115
better reward
better reward
better reward
score: 161
better reward
score: 167
score: 19
score: 103
better reward
better reward
better reward
better reward
score: 160
score: 47
better reward
better reward
better reward
better reward
score: 186
score: 69
score: 117
score: 89
better reward
better reward
better reward
score: 155
score: 179
score: 118
score: 151
score: 92
score: 103
score: 98
score: 53
score: 99
score: 133
score: 85
score: 63
score: 87
score: 128
score: 93
score: 67
score: 119
score: 105
better reward
score: 82
score: 21
score: 18
score: 168
score: 85
score: 89
score: 127
score: 101
score: 83
better reward
score: 103
score: 86
b

KeyboardInterrupt: 

In [378]:
save_model('1dec-model-both-v1', brain_array)
print(len(brain_array))

121466


So I trained this model for hours. it now has 121466 states in its memory. Let's see how well it performs

In [393]:
brain_array = get_model('1dec-model-both-v1')

env = gym.make("CartPole-v1")
env = gym.make("CartPole-v1", render_mode="human") # uncomment this to actually see the progress (slows down the process terribly)

scores_array = []
MAX_AMOUNT_OF_GAMES = 3
MAX_AMOUNT_OF_FRAMES_PER_GAME = 500
TRY_IMPROVE_REWARD = 0
Test()
print('average score: ' + str(get_avg_score(MAX_AMOUNT_OF_GAMES)))
print(len(brain_array))

score: 156
score: 97
score: 121
average score: 124.66666666666667
121466


we got an average score of 136.17, that's again an improvement on the last time. Let's add it to the scoreboard

Scoreboard
1. 1dec-model-both-v1: 136.17 (121466) (angle+speed)
2. 1dec-model-angle-v1: 107.00 (13121) (angle method) (got lost in the proces, oops)
3. 1dec-model-v3: 99.89 (11846) (speed method)
4. Random Agent: 23.4
5. 1dec-model-v2: 19.08  (3050) (speed method)

If i observe the current gameplay of the model I see something i expected. The model now learned how to slowly keep moving to one direction, but this won't work because it will hit a wall at some point. The most clean way to fix this would be to let the model look into the future, and base rewards of of looking multiple steps into the future. This is definitely not impossible. I might consider trying it later if I have time left.